alphametics/example.go

```diff
@@ -0,0 +1,200 @@
+package alphametics
+
+import "sync"
+
+const TestVersion = 1
+
+// Set this to true to debug the example. It enables several useful logging tools.
+const debugging = false
+
+type accumulator struct {
+	carry    int
+	usedNums uint16       // bit n denotes if digit n is taken
+	solution map[rune]int // partially filled solution map
+}
+
+type ipc struct {
+	resultChan chan map[rune]int
+	abortChan  chan struct{}
+	waitGroup  *sync.WaitGroup
+}
+
+func Solve(a, b, c string) map[rune]int {
+	aRunes, bRunes, cRunes := prepareRunes(a, b, c)
+
+	// Setup the machinery for solving in parallel.
+	resultChan := make(chan map[rune]int)
+	abortChan := make(chan struct{}) // Closed when goroutines can abort
+
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(1)
+	go func() {
+		// This goroutine ensures the main thread will return nil if no answer can
+		// be found instead of blocking forever.
+		waitGroup.Wait()
+		resultChan <- nil
+	}()
+	go solveStep(aRunes, bRunes, cRunes, accumulator{}, &ipc{resultChan, abortChan, &waitGroup})
+
+	result := <-resultChan
+	close(abortChan)
+	return result
+}
+
+func solveStep(aRunes, bRunes, cRunes []rune, acc accumulator, ipc *ipc) {
+	defer ipc.waitGroup.Done()
+	if abortSignalled(ipc) {
+		return
+	}
+	aRune, bRune, cRune := aRunes[0], bRunes[0], cRunes[0]
+	c, cFound := acc.solution[cRune]
+	for _, a := range availableDigits(aRune, acc, 0, 0) {
+		for _, b := range availableDigits(bRune, acc, aRune, a) {
```

Comment:
Interestingly, this looks like a case for recursion (which you are doing below), but are not doing here.